src/cluster.rs

```diff
@@ -433,6 +440,114 @@ impl ClusterConnection {
             }
         }
     }
+
+    fn send_recv_and_retry_cmds(&self, cmds: &[Cmd]) -> RedisResult<Vec<Value>> {
+        // Vector to hold the results, pre-populated with `Nil` values. This allows the original
+        // cmd ordering to be re-established by inserting the response directly into the result
+        // vector (e.g., results[10] = response).
+        let mut results: Vec<Value> = Vec::with_capacity(cmds.len());
+        results.resize(cmds.len(), Value::Nil);
+
+        let to_retry = self
+            .send_all_commands(&cmds)
+            .and_then(|node_cmds| self.recv_all_commands(&mut results, &node_cmds))?;
+
+        if to_retry.is_empty() {
+            return Ok(results);
+        }
+
+        // Refresh the slots to ensure that we have a clean slate for the retry attempts.
+        self.refresh_slots()?;
+
+        // Given that there are commands that need to be retried, it means something in the cluster
+        // topology changed. Execute each command seperately to take advantage of the existing
+        // retry logic that handles these cases.
+        for retry_idx in to_retry {
+            let cmd = &cmds[retry_idx];
+            results[retry_idx] = self.request(cmd, move |conn| conn.req_command(cmd))?;
+        }
+        Ok(results)
+    }
+
+    // Build up a pipeline per node, then send it
+    fn send_all_commands(&self, cmds: &[Cmd]) -> RedisResult<Vec<NodeCmd>> {
+        let mut connections = self.connections.borrow_mut();
+
+        let node_cmds = self.map_cmds_to_nodes(cmds)?;
+        for nc in &node_cmds {
+            self.get_connection_by_addr(&mut connections, &nc.addr)?
+                .send_packed_command(&nc.pipe)?;
+        }
+        Ok(node_cmds)
+    }
+
+    fn get_addr_for_cmd(&self, cmd: &Cmd) -> RedisResult<String> {
+        let slots = self.slots.borrow();
+
+        let addr_for_slot = |slot: u16| -> RedisResult<String> {
+            let (_, addr) = slots
+                .range(&slot..)
+                .next()
+                .ok_or((ErrorKind::ClusterDown, "Missing slot coverage"))?;
+            Ok(addr.to_string())
+        };
+
+        match RoutingInfo::for_routable(cmd) {
+            Some(RoutingInfo::Random) => {
+                let mut rng = thread_rng();
+                Ok(addr_for_slot(rng.gen::<u16>() % SLOT_SIZE as u16)?)
+            }
+            Some(RoutingInfo::Slot(slot)) => Ok(addr_for_slot(slot)?),
+            _ => fail!(UNROUTABLE_ERROR),
+        }
+    }
+
+    fn map_cmds_to_nodes(&self, cmds: &[Cmd]) -> RedisResult<Vec<NodeCmd>> {
+        let mut cmd_map: HashMap<String, NodeCmd> = HashMap::new();
+
+        for (idx, cmd) in cmds.iter().enumerate() {
+            let addr = self.get_addr_for_cmd(&cmd)?;
+            let nc = cmd_map
+                .entry(addr.clone())
+                .or_insert_with(|| NodeCmd::new(addr));
+            nc.indexes.push(idx);
+            nc.pipe.extend(cmd.get_packed_command());
```

Comment:
```suggestion
            cmd.write_packed_command(&mut nc.pipe);
```
Avoids redundant `Vec` allocations